[DOC] Add script processor documentation (#5988)
opensearch-project/documentation-website: _ingest-pipelines/processors/script.md

---
layout: default
title: script
parent: Ingest processors
nav_order: 230
---

# `script` processor

> The `script` processor executes inline and stored scripts that can modify or transform data in an OpenSearch document during the ingestion process. The processor uses script caching for improved performance, as scripts may be recompiled per document. Refer to [Script APIs](https://opensearch.org/docs/latest/api-reference/script-apis/index/) for information about working with scripts in OpenSearch. 

Reviewer: ```suggestion
The `script` processor executes inline and stored scripts that can modify or transform data in an OpenSearch document during the ingestion process. The processor uses script caching for improved performance because scripts may be recompiled per document. Refer to [Script APIs](https://opensearch.org/docs/latest/api-reference/script-apis/index/) for information about working with scripts in OpenSearch. 
```

The following is the syntax for the `script` processor:

```json
{
  "processor": {
    "script": {
      "source": "<script_source>",
      "lang": "<script_language>",
      "params": {
        "<param_name>": "<param_value>"
      }
    }
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `script` processor.

| Parameter  | Required/Optional  | Description  |
|---|---|---|

> `source`  | Optional  | The Painless script to be executed. Either `id` or `source` must be specified, but not both. If `source` is provided, then the script is executed from the provided source code.

Reviewer: ```suggestion
`source`  | Optional  | The Painless script to be executed. Either `id` or `source` must be specified---but not both. If `source` is specified, then the script is executed using the provided source code.
```

> `id` | Optional | The ID of a stored script that was previously created using the [Create Stored Script API]({{site.url}}{{site.baseurl}}/api-reference/script-apis/create-stored-script/). Either `id` or `source` must be specified, but not both. If `id` is provided, then the script source is retrieved from the stored script with the specified ID. 

Reviewer: ```suggestion
`id` | Optional | The ID of a stored script previously created using the [Create Stored Script API]({{site.url}}{{site.baseurl}}/api-reference/script-apis/create-stored-script/). Either `id` or `source` must be specified, but not both. If `id` is specified, then the script source is retrieved from the stored script with the specified ID. 
```

`lang`  | Optional  | The programming language of the script. Default is `painless`.
`params` | Optional |  The parameters that can be passed to the script.
`description`  | Optional  | A description of the processor's purpose or configuration.
`if` | Optional | Specifies to conditionally execute the processor.

> `ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer: ```suggestion
`ignore_failure` | Optional | Specifies to ignore processor failures. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

        "lang": "painless",
        "description": "Convert message field to uppercase"
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Step 2 (Optional): Test the pipeline

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json
POST _ingest/pipeline/my-script-pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "message": "hello, world!"
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

The following example response confirms that the pipeline is working as expected:

```json
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "message": "HELLO, WORLD!"
        },
        "_ingest": {
          "timestamp": "2024-05-30T16:24:23.30265405Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Step 3: Ingest a document 

The following query ingests a document into an index named `testindex1`:

```json
POST testindex1/_doc?pipeline=my-script-pipeline
{
  "message": "hello, world!"
}
```
{% include copy-curl.html %}

#### Response

> This response confirms that the document is indexed into `testindex1` and is indexing all documents with the `message` field converted to uppercase.

Reviewer: ```suggestion
The response confirms that the document has been indexed into `testindex1` and is indexing all documents with the `message` field converted to uppercase:
```

> This response confirms that the document is indexed into `testindex1` and is indexing all documents with the `message` field converted to uppercase.

Reviewer: "is indexing" => either "has indexed" or "will index"

Author: Revised "has indexed"

---
layout: default
title: script
parent: Ingest processors
nav_order: 230
---

# `script` processor

The `script` processor executes inline and stored scripts that can modify or transform data in an OpenSearch document during the ingestion process. The processor uses script caching for improved performance, as scripts may be recompiled per document. Refer to [Script APIs](https://opensearch.org/docs/latest/api-reference/script-apis/index/) for information about working with scripts in OpenSearch. 

The following is the syntax for the `script` processor:

```json
{
  "processor": {
    "script": {

>       "tag": "<processor_tag>",

Reviewer 2: `tag` and `description` are common parameters, we may remove them here to make it consistent with the documentation of other processors which don't have these two parameters.

      "description": "<processor_description>",
      "source": "<script_source>",
      "lang": "<script_language>",
      "params": {
        "<param_name>": "<param_value>"
      }
    }
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `script` processor.

| Parameter  | Required/Optional  | Description  |
|---|---|---|

> `source`  | Required  | The Painless script to be executed.

Reviewer 2: `source`, `lang` and `params` are optional, and another optional parameter `id` is missing, `id` can be used to reference a stored script which was created by this API:https://opensearch.org/docs/latest/api-reference/script-apis/create-stored-script/, `source` and `id` are mutually exclusive, either of them can be set.

`lang`  | Required  | The programming language of the script. Default is `painless`.
`params` | Required |  The parameters that can be passed to the script.
`description`  | Optional  | A description of the processor's purpose or configuration.
`if` | Optional | Specifies to conditionally execute the processor.
`ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`on_failure` | Optional | Specifies a list of processors to run if the processor fails during execution. These processors are executed in the order they are specified.
`tag` | Optional | An identifier tag for the processor. Useful for debugging in order to distinguish between processors of the same type.

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline named <pipeline name> that uses the script processor to <do what?>: 

```json

> PUT _ingest/pipeline/my-script-pipeline

Reviewer 2: This example looks good to me.

---
layout: default
title: Script
parent: Ingest processors
nav_order: 230
---

# Script processor

The `script` processor is used to <explain what is used to do>.

The following is the syntax for the `script` processor:

```json
<insert syntax example>

> ```

Author: @@gaobinlong Is this syntax correct?

{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `script` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
<insert the parameters>

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline, named <name>, that uses the `script` processor to <do what?>: 

> ```json

Author: @gaobinlong Will you provide an example query?
